### PR TITLE
Mushroom classifier train, test, and infer methods completed

### PR DESCRIPTION
**Pull request by sram-alectio**
I haven't got the end-to-end test to work yet, but just pushing the completed code for process.py

### CODE REVIEW

File: alectio_sdk/flask_wrapper/pipeline.py [hunk 1]
```diff
@@ -102,7 +102,7 @@ def _one_loop(self, payload, args):
         # self.curout_loop = payload["cur_loop"]
 
         # Leave cur_loop - 1 when doing a 1 dag solution, when doing 2 dag cur_loop remains the same
-        self.cur_loop = payload["cur_loop"] - 1
+        self.cur_loop = payload["cur_loop"]
         self.bucket_name = payload["bucket_name"]
 
         # type of the ML problem
```
Comment (AntonMu) [inline alectio_sdk/flask_wrapper/pipeline.py:108]: Please don't change the underlying code base when doing an example. If you need changes, you have to do it in your experiment.

Changing this would break other working experiments.

File: alectio_sdk/flask_wrapper/pipeline.py [hunk 2]
```diff
@@ -142,7 +142,7 @@ def _one_loop(self, payload, args):
             object_key = os.path.join(self.expt_dir, "data_map.pkl")
             self.client.multi_part_upload_with_s3(self.state_json, self.bucket_name, object_key, "pickle")
         else:
-            self.resume_from = "ckpt_{}".format(self.cur_loop - 1)
+            self.resume_from = "ckpt_{}".format(self.cur_loop)
```
Comment (AntonMu) [inline alectio_sdk/flask_wrapper/pipeline.py:145]: same as above

File: examples/mushroom_classification/processes.py [hunk 3]
```diff
@@ -1,32 +1,51 @@
...
+import argparse
 
 device = "cuda" if torch.cuda.is_available() else "cpu"
 
-def getdatasetstate(args={}):
-    return {k: k for k in range(120000)}
+print("using device", device)
+
+def getdatasetstate(args, split="train"):
+    global train_dataset, test_dataset
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:20]: don't use globals - these functions needs to be flexible so the SDK can use them. Since this is tabular data, just put 
```
def getdatasetstate(config_args={}):  
    return {k: k for k in range(config_args["trainsize"])}
```

File: examples/mushroom_classification/processes.py [hunk 4]
```diff
@@ -1,32 +1,51 @@
...
+    ix = 0
+    for row in dataset[:][0]:
+        referencedict[ix] = row 
+        ix += 1                           
+        
+    return referencedict
 
 
 def train(args, labeled, resume_from, ckpt_file):
+    print("running training")
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:48]: no print

File: examples/mushroom_classification/processes.py [hunk 5]
```diff
@@ -1,32 +1,51 @@
...
     batch_size = args["batch_size"]
     lr = 1e-4
     momentum = 0.9
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:51]: Hyperparemeters should be set in the config file.

File: examples/mushroom_classification/processes.py [hunk 6]
```diff
@@ -42,13 +61,11 @@ def train(args, labeled, resume_from, ckpt_file):
     train_dataset = train_test[0]
     test_dataset = train_test[1]
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:63]: no globals - load them from file by using args.

File: examples/mushroom_classification/processes.py [hunk 7]
```diff
@@ -42,13 +61,11 @@ def train(args, labeled, resume_from, ckpt_file):
...
-    #print("Test set length:", len(test))
-
     train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
     test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=True)
     
-    net = NeuralNet().to(device)
+    net = NeuralNet()
+    net = net.to(device=device)
 
     print(net)
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:70]: no print

File: examples/mushroom_classification/processes.py [hunk 8]
```diff
@@ -85,73 +106,140 @@ def train(args, labeled, resume_from, ckpt_file):
     print("done training")
 
     print("Finished Training. Saving the model as {}".format(ckpt_file))
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:108]: no print

File: examples/mushroom_classification/processes.py [hunk 9]
```diff
@@ -85,73 +106,140 @@ def train(args, labeled, resume_from, ckpt_file):
...
-    print("Training accuracy: {}".format((train_acc / len(chosen_train_dataset) * 100)))
+
     ckpt = {"model": net.state_dict(), "optimizer": optimizer.state_dict()}
-    torch.save(ckpt, os.path.join(args["EXPT_DIR"], ckpt_file))
+    #print("ckpt file", ckpt)
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:111]: remove

File: examples/mushroom_classification/processes.py [hunk 10]
```diff
@@ -85,73 +106,140 @@ def train(args, labeled, resume_from, ckpt_file):
...
+    torch.save(ckpt, os.path.join(args["EXPT_DIR"], ckpt_file + ".pth"))
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:112]: try to make this more robust - right now it depends on the location you call this function from. But you want to use absolute paths instead.

File: examples/mushroom_classification/processes.py [hunk 11]
```diff
@@ -85,73 +106,140 @@ def train(args, labeled, resume_from, ckpt_file):
...
 
     return
 
 
 def test(args, ckpt_file):
+    print("running testing")
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:118]: no print

File: examples/mushroom_classification/processes.py [hunk 12]
```diff
@@ -85,73 +106,140 @@ def train(args, labeled, resume_from, ckpt_file):
...
     batch_size = args["batch_size"]
-    testloader = DataLoader(test_dataset, batch_size=batch_size, shuffle=False, collate_fn=generate_batch) 
+    lr = 1e-4
+    momentum = 0.9
+    epochs = args["train_epochs"]
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:123]: see above

File: examples/mushroom_classification/processes.py [hunk 13]
```diff
@@ -85,73 +106,140 @@ def train(args, labeled, resume_from, ckpt_file):
...
 
-    predictions, targets = [], []
-    net = TextSentiment(VOCAB_SIZE, EMBED_DIM, NUN_CLASS).to(device)
-    ckpt = torch.load(os.path.join(args["EXPT_DIR"], ckpt_file))
-    net.load_state_dict(ckpt["model"])
-    net.eval()
+    global train_dataset, test_dataset
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:124]: no globals

File: .gitignore [hunk 14]
```diff
@@ -156,6 +156,7 @@ cython_debug/
 data
 log
 vector
+FaceBoxes.PyTorch/
```
Comment (AntonMu) [inline .gitignore:159]: Only add this to the .gitignore within the examples. Also for the facebox demo, it is better to make a new feature branch.

File: examples/mushroom_classification/processes.py [hunk 15]
```diff
@@ -1,56 +1,42 @@
...
-    test_dataset = train_test[1]
+    train_subset = Subset(train_dataset, labeled)
 
-    #print("Train set length:", len(train))
-    #print("Test set length:", len(test))
+    print("The length of the train subset is", len(train_subset))
 
-    train_loader = DataLoader(train_dataset, batch_size=batch_size, shuffle=True)
-    test_loader = DataLoader(test_dataset, batch_size=batch_size, shuffle=True)
+    train_loader = DataLoader(train_subset, batch_size=batch_size, shuffle=False)
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:36]: `shuffle=True` it's better to shuffle between epochs.

File: examples/mushroom_classification/processes.py [hunk 16]
```diff
@@ -81,77 +71,141 @@ def train(args, labeled, resume_from, ckpt_file):
...
+                batch_y = batch_y.to(device)
+                prediction = net(batch_x)
+                
+            for logit in prediction:
+                predictions[predix] = logit.cpu().numpy().tolist()
+                predix+=1
+            
+            pbar.update()
+    
+    return {"outputs": predictions}
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:183]: You still need to make sure to keep track of your unlabeled indices in the output. 

For examples if unlabeled is `[5,8,9]` your output should look like `{5: [.4,.6], 8: [.7,.3], 9: [.2, .8]}`

File: examples/mushroom_classification/processes.py (hunk 10 above)
Comment (AntonMu) [reply examples/mushroom_classification/processes.py:77]: This still needs to be addressed.

File: examples/mushroom_classification/processes.py (hunk 16 above)
Comment (AntonMu) [reply examples/mushroom_classification/processes.py:183]: This still needs to be addressed. To debug, look at the outputs of infer(...) - is it what you expect?

File: examples/mushroom_classification/processes.py [hunk 17]
```diff
@@ -82,76 +71,137 @@ def train(args, labeled, resume_from, ckpt_file):
...
     ckpt_file = "ckpt_0"
 
-    train(labeled=labeled, resume_from=resume_from, ckpt_file=ckpt_file)
-    test(ckpt_file=ckpt_file)
-    infer(unlabeled=[10, 20, 30], ckpt_file=ckpt_file)
+    print("Testing getdatasetstate")
+    getdatasetstate(args=args)
+    train(args=args, labeled=labeled, resume_from=resume_from, ckpt_file=ckpt_file)
+    test(args=args, ckpt_file=ckpt_file)
+    infer(args=args, unlabeled=[10, 20, 30], ckpt_file=ckpt_file)
```
Comment (AntonMu) [inline examples/mushroom_classification/processes.py:207]: `print(infer(args=args, unlabeled=[10, 20, 30], ckpt_file=ckpt_file))`

File: examples/mushroom_classification/requirements.txt [hunk 18]
```diff
@@ -0,0 +1,60 @@
+alectio-sdk==0.0.1
```
Comment (AntonMu) [inline examples/mushroom_classification/requirements.txt:1]: Have you tested this line?

File: examples/mushroom_classification/processes.py (hunk 15 above)
Comment (AntonMu) [reply examples/mushroom_classification/processes.py]: This comment still needs to be addressed

File: examples/mushroom_classification/requirements.txt (hunk 18 above)
Comment (sram-alectio) [reply examples/mushroom_classification/requirements.txt:1]: Yes I have.

Comment (AntonMu) [reply examples/mushroom_classification/requirements.txt:1]: image

Comment (sram-alectio) [reply examples/mushroom_classification/requirements.txt:1]: image
I did not have this issue while installing, however for clarity I have removed the alectio-sdk from the requirements.txt file and have added an instruction for users to install it manually using `pip install ../../.` in the readme.

File: examples/mushroom_classification/processes.py (hunk 17 above)
Comment (sram-alectio) [reply examples/mushroom_classification/processes.py:207]: done.

Comment (sram-alectio) [reply examples/mushroom_classification/processes.py:207]: also added the current working directory in torch.save() - to address your comments above.
